eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,201 @@
+            // if not exist in the classPathMap, that implies the type is java default types, such as int.
+            if (classPathMap.containsKey(returnTypeName)) {
+                if (!isValidLibrary(classPathMap.get(returnTypeName))) {
+                    log(typeToken, String.format(EXTERNAL_DEPENDENCY_ERROR, returnTypeName));
+                }
+            }
+        } else {
+            checkTypeArguments(typeToken);
+        }
+    }
+
+    /**
+     * Checks for input parameters
+     *
+     * @param parametersTypeToken PARAMETERS AST node
+     */
+    private void checkParametersType(DetailAST parametersTypeToken) {
+        for (DetailAST ast = parametersTypeToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+            if (ast.getType() == TokenTypes.PARAMETER_DEF) {
+                checkTypeArguments(ast.findFirstToken(TokenTypes.TYPE));
+            }
+        }
+    }
+
+    /**
+     *  A helper function that checks TYPE AST node. Since both return type and input parameter argument type has
+     *  TYPE AST node under. This function applied to both.
+     *
+     * @param typeToken TYPE AST node
+     */
+    private void checkTypeArguments(DetailAST typeToken) {
+        final DetailAST identToken = typeToken.findFirstToken(TokenTypes.IDENT);
+        // if there is no IDENT token, implies the token is default java types.
+        if (identToken != null) {
+            final String typeName = identToken.getText();
+            // if not exist in the classPathMap, that implies the type is java default types, such as int.
+            if (classPathMap.containsKey(typeName)) {
+                if (!isValidLibrary(classPathMap.get(typeName))) {
+                    log(typeToken, String.format(EXTERNAL_DEPENDENCY_ERROR, typeName));
+                } else {
+                    // Checks multiple type arguments
+                    final DetailAST typeArgumentsToken = typeToken.findFirstToken(TokenTypes.TYPE_ARGUMENTS);
+                    if (typeArgumentsToken != null) {
+                        for (DetailAST ast = typeArgumentsToken.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+                            if (ast.getType() == TokenTypes.TYPE_ARGUMENT) {
+                                checkSingleTypeArgument(ast);
+                            }
+                        }
+                    }
+                    // Checks single type argument
+                    final DetailAST typeArgumentToken = typeToken.findFirstToken(TokenTypes.TYPE_ARGUMENT);
+                    if (typeArgumentToken != null) {
+                        checkSingleTypeArgument(typeArgumentToken);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     *  A helper function that checks for single type arguments scenario.
+     *
+     * @param typeArgumentToken TYPE_ARGUMENT AST node
+     */
+    private void checkSingleTypeArgument(DetailAST typeArgumentToken) {
```

Comment:
Tbh, I am not a fan of methods that have side-effects in addition to performing work. It makes it harder to write unit tests for these sorts of methods.

```
private Result checkArgumentType(DetailAST argument) {
...
}

private static class Result {
    boolean isValid();
    // Name of all the types that expose an external dependency.
    String[] externalDependencies();
}
```